Lib/fontTools/ttLib/tables/_g_l_y_f.py

```diff
@@ -1080,20 +1145,25 @@ def recalcBounds(self, glyfTable):
         recomputed when the ``coordinates`` change. The ``table__g_l_y_f`` bounds
         must be provided to resolve component bounds.
         """
-        coords, endPts, flags = self.getCoordinates(glyfTable)
-        self.xMin, self.yMin, self.xMax, self.yMax = calcIntBounds(coords)
+        try:
+            coords, endPts, flags = self.getCoordinates(glyfTable)
+            self.xMin, self.yMin, self.xMax, self.yMax = calcIntBounds(coords)
+        except NotImplementedError:
```

Comment:
The `pass` here causes the old xMin, yMin etc. to be left alone, which makes messes up the lsb calculation in ttGlyphSet. I *think* you work around this by setting offset to zero in there's a location stack, but I fear the "normal" drawing (not as part of a composite) of such a glyph is still messed up.

Reply:
My understanding was that offset doesn't apply to components. But now that I come to think about it, I'm probably wrong here. Ugh. This is really messy :(.

Reply:
In HarfBuzz we only apply the shift to toplevel glyph. That's what tripped me. I'm utterly confused now. Let me study fonttools code a bit, and possibly construct a font.

Reply:
Sorry, I'm also confused: I'm observing things in my code that do not happen in your code, and it relates to the offset. My conclusion above may be wrong. I'll debug further.

Reply:
So indeed HarfBuzz and FontTools have different behavior in this regard currently.  I subset MutatorSans to have the Composite glyph Á. Then I changed the component accent glyph's lsb in the hmtx to shift it horizontall. HarfBuzz still renders the Á unchanged, while FontTools render it with the accent shifted.

I need volunteers to test this in Windows / Mac. The font is attached.
[comp.ttf.gz](https://github.com/fonttools/fonttools/files/10581839/comp.ttf.gz)

Reply:
> I need volunteers to test this in Windows / Mac. The font is attached.

Mac shows the accent shifted. So HarfBuzz is in the wrong *compared to Mac*. We need Windows testing as well.

Reply:
> Mac shows the accent shifted. So HarfBuzz is in the wrong compared to Mac. We need Windows testing as well.

FreeType agrees with HarfBuzz though.

Reply:
For reference:

HarfBuzz and FreeType rendering (applying offset only at top-level; not applying to components):
![image](https://user-images.githubusercontent.com/142891/216684601-22c4e7bd-607a-443a-b21b-d4c9360e4170.png)

Reply:
Mac rendering:
<img width="96" alt="Screenshot 2023-02-03 at 12 00 30 PM" src="https://user-images.githubusercontent.com/142891/216685372-a089ff18-65ac-48c6-bb5d-93a01971641a.png">

Reply:
FontTools rendering:
![image](https://user-images.githubusercontent.com/142891/216685633-acbd8e15-85c0-4bcb-9fd0-269fd4a74dcf.png)

Reply:
Windows rendering (different font but same hack):
![image](https://user-images.githubusercontent.com/142891/216689055-b3c71848-52ee-4a1b-a998-6fcb84459fd6.png)

So, looks like Mac is in the minority here and FontTools follows it. @nedley

Reply:
The implications for FontTools is that `DecomposingPen` cannot work as it does, and the Composite decomposing should also be moved to `ttGlyphSet`. @justvanrossum

Reply:
@nedley The issue we are discussing is that if a component glyph in a glyf table has different xMin and hmtx lsb, Windows and FreeType do _not_ shift it by the difference when using the component, whereas CoreText seems to shift.

Ie. Windows and FreeType only shift top-level glyphs by the difference of xMin/lsb.

Reply:
@justvanrossum In fact, fontTools itself has internal inconsistency in that `decomposingPen` applies offset to components, but `_g_l_y_f.Glyph.getCoordinates` doesn't!

Reply:
We should probably open a new issue for this.

The fact that drawing components should not apply the offset is new to me. How is this written in the spec?

It indeed makes the current way to decompose things impossible to handle correctly for non-zero offsets.

That said: I'm still confused. Do I understand correctly that the lsb in hmtx should make no difference for glyphs used as (var) components?

Yet, if I hack up Tests/ttLib/data/varc-ac00-ac01.ttf to have different lsb values for glyph00003, glyph00004 and glyph00005, I don't get a correct rendered uniAC00. I tested with
```
python -m fontTools.pens.svgPathPen Tests/ttLib/data/varc-ac00-ac01.ttf 가 > test.svg
```
And
```
python -m fontTools.pens.svgPathPen varc-ac00-ac01-lsb-hack.ttf 가 > test-lsb-hack.svg
```
with the attached font.
[varc-ac00-ac01-lsb-hack.ttf.zip](https://github.com/fonttools/fonttools/files/10582330/varc-ac00-ac01-lsb-hack.ttf.zip)

Perhaps I am misunderstanding things.

Reply:
> We should probably open a new issue for this.

Agreed. Will do.

> The fact that drawing components should not apply the offset is new to me. How is this written in the spec?

The HarfBuzz code calls the whole offset application "undocumented behavior".   Indeed, I cannot find anywhere in the spec that says glyph shape should be shifted to align xMin to lsb.

> It indeed makes the current way to decompose things impossible to handle correctly for non-zero offsets.

Right.

> That said: I'm still confused. Do I understand correctly that the lsb in hmtx should make no difference for glyphs used as (var) components?

That's what Windows and FreeType seem to be doing.

> Yet, if I hack up Tests/ttLib/data/varc-ac00-ac01.ttf to have different lsb values for glyph00003, glyph00004 and glyph00005, I don't get a correct rendered uniAC00.

Humm. Let me try. That's exactly why I added the `offset = 0` change.

Reply:
I think in general, it is undesirable to have lsb to be not equal to xMin. fontmake generally doesn't (can't?) produce fonts like that, and I'm pretty sure neither does Glyphs.

I would be fine with fonttools saying: "lsb != xMin is not supported in the glyphset draw API for glyphs that are used as components".

The fact that you only now discovered a mismatch between Windows and  macOS says a lot about how uncommon this is in real-world fonts.

That said, I was hoping to fix the problem I am seeing on my end (with hacked up Fontra code to use addVarComponent to extract variable components from .ttf) by fixing the test font to have lsb == xMin in all glyphs. And that didn't work, so I must look further. Unless something is fishy in the ttGlyphSet implementation, which is entirely possible, given how incredibly hairy this offset stuff is.

Reply:
> Yet, if I hack up Tests/ttLib/data/varc-ac00-ac01.ttf to have different lsb values for glyph00003, glyph00004 and glyph00005, I don't get a correct rendered uniAC00.

Something broken in my `fromXML` impl I think. If I diff the two fonts, all the `translateX` and `translateY` are gone. That's a relief. Debugging.

Reply:
> Something broken in my fromXML impl I think. If I diff the two fonts, all the translateX and translateY are gone. That's a relief. Debugging.

The compile code was broken. Fixed. With that I recreated your work and it renders fine still; with FontTools and HarfBuzz.

Reply:
> Indeed, I cannot find anywhere in the spec that says glyph shape should be shifted to align xMin to lsb.

The only thing I find is this, which is rather indirect:

> Note: The scaler will perform better if the glyph coordinates have been created such that the xMin is equal to the lsb. For example, if the lsb is 123, then xMin for the glyph should be 123. If the lsb is -12 then the xMin should be -12. If the lsb is 0 then xMin is 0. If all glyphs are done like this, set bit 1 of flags field in the 'head' table.
 
https://learn.microsoft.com/en-us/typography/opentype/spec/glyf#glyph-headers

And nothing about components indeed.

Reply:
> The pass here causes the old xMin, yMin etc. to be left alone

so, what do we expect recalcBounds to do with bbox when a glyph contain varcomponents?

Reply:
I think we have to use ttGlyphSet and ControlBoundsPen and do it the slow but correct way. There may be some chicken/egg problem involved, as that drawing code also calls recalcBounds.

However: I don't think the drawing code *should* call recalcBounds, because the only reason it does so is to calculate the offset, in case xMin is different from lsb. But: we do not have to cater for that situation, as lsb *has* to be equal to xMin for variable fonts.

I'm afraid it may take some refactoring to get this right. Perhaps postpone until after this PR lands?